bootloader/src/pyi_launch.c

```diff
@@ -390,53 +385,48 @@ pyi_launch_run_scripts(ARCHIVE_STATUS *status)
         if (ptoc->typcd == ARCHIVE_ITEM_PYSOURCE) {
             /* Get data out of the archive.  */
             data = pyi_arch_extract(status, ptoc);
-            /* Set the __file__ attribute within the __main__ module,
-             *  for full compatibility with normal execution. */
-            namelen = strnlen(ptoc->name, PATH_MAX);
-            if (namelen >= PATH_MAX-strlen(".py")-1) {
-                FATALERROR("Name exceeds PATH_MAX\n");
-                return -1;
-            }
-
-            strcpy(buf, ptoc->name);
-            strcat(buf, ".py");
-            VS("LOADER: Running %s\n", buf);
-
-            if (is_py2) {
-                __file__ = PI_PyString_FromString(buf);
-            }
-            else {
-                __file__ = PI_PyUnicode_FromString(buf);
-            };
-            PI_PyObject_SetAttrString(__main__, "__file__", __file__);
-            Py_DECREF(__file__);
+            VS("LOADER: Running %s.py\n", ptoc->name);
 
             /* Unmarshall code object */
             code = PI_PyMarshal_ReadObjectFromString((const char *) data, ntohl(ptoc->ulen));
-
             if (!code) {
                 FATALERROR("Failed to unmarshal code object for %s\n", ptoc->name);
                 PI_PyErr_Print();
                 return -1;
             }
-            /* Run it */
-            retval = PI_PyEval_EvalCode(code, main_dict, main_dict);
+
+            /* Run the code */
+            /* For full compatibility with normal execution, the __file__
+             * attribute within the __main__ module needs to be set.
+             * PyImport_ExecCodeModule() will set it based on
+             * code.co_filename. PyInstaller takes care that co_filename is
+             * correct. */
+            module = PI_PyImport_ExecCodeModule("__main__", code);
 
             /* If retval is NULL, an error occured. Otherwise, it is a Python object.
              * (Since we evaluate module-level code, which is not allowed to return an
              * object, the Python object returned is always None.) */
-            if (!retval) {
+            if (!module) {
                 PI_PyErr_Print();
                 /* If the error was SystemExit, PyErr_Print calls exit() without
                  * returning. So don't print "Failed to execute" on SystemExit. */
                 FATALERROR("Failed to execute script %s\n", ptoc->name);
                 return -1;
             }
+            Py_DECREF(module);
```

Comment:
Wait, how can you decref the module and then complain when you can no longer access its attributes?

Reply:
I would expect the code to reference the dict, too. But anyway, I also tried *not* decrementing the reference count here with no success. But I'll double-check.